### DISCUSSION TITLE
Is there a plan to output in 'standalone' mode?

### DISCUSSION THREAD

**Discussion by yyong008** [remix-run/remix#9460] [General]
When deploying some larger projects using Docker, it might be more convenient.

**Comment by sergiodxa** [comment]
What do you mean "standalone mode"? You can run a Remix app inside a Docker container the same way as any Node app, just run `npm start`.

**Reply by yyong008** [reply]
The Vite build outputs two directories: client and server, both of which depend on node_modules. The standalone output should relate solely to the production environment and include the node_modules directory.

**Reply by sergiodxa** [reply]
You mean you want to bundle node_modules? That's something you can configure in Vite itself, previously it was an option in `remix.config` to bundle the server dependencies.

Also `build/client` doesn't really need node_modules, since that's the client-side JS it doesn't depend on anything from node_modules, that's only used server-side to get better stack traces.

Without configuring Vite to bundle everything server-side, the recommendation is to first install the dependencies, then run the build and after that run `npm install --production` to only keep production dependencies (and ensure only the dependencies you need at runtime as marked as `dependencies` and the rest as `devDependencies`), with this your Docker image will have the `build`, `public`, and `node_modules` folders only, and the package.json to run `npm start`.

**Reply by yyong008** [reply]
sometimes, even when configuring external dependencies for some libraries in Vite, dependencies from node_modules are still needed.

**Comment by SteveSuv** [comment]
you can use `tsup` to build `build/server/index.js` again and pack all dependencies to a single js file.
```ts
// tsup.config.ts
{
    target: 'node16',
    platform: 'node',
    noExternal: [/(.*)/],
    ...rest of options
}
```
https://github.com/egoist/tsup/issues/619#issuecomment-1420423401

**Reply by yyong008** [reply]
I will be try
